File: go/05_array/array.go
package array

import (
	"errors"
	"fmt"
)

//Array 声明一个Array结构体
type Array struct {
	data   []int
	length uint
}

/*
*数组的插入，查找，越界，删除等操作
 */

//NewArray 初始化数组内存
func NewArray(capacity uint) *Array {
	if capacity == 0 {
		return nil
	}
	return &Array{
		data:   make([]int, capacity, capacity),
		length: 0,
	}
}

//获取数组长度
func (this *Array) Len() uint {
	return this.length
}

//判断索引是否越界
func (this *Array) IsIndexOutOfRange(index uint) bool {
	if index >= uint(cap(this.data)) { //超过数组最大容量，越界了
		return true
	}
	return false
}

//通过索引查找数组
func (this *Array) Find(index uint) (int, error) {
	if this.IsIndexOutOfRange(index) {
		return 0, errors.New("out of range")
	}
	return this.data[index], nil
}

//插入数值到索引指定位置
func (this *Array) Insert(index uint, val int) error {
	if this.Len() == uint(cap(this.data)) { //数组已满
		return errors.New("full array")
	}

	if index != this.length && this.IsIndexOutOfRange(index) {
		return errors.New("out of range")
	}

	for i := this.length; i > index; i-- {
		this.data[i] = this.data[i-1]
	}

	this.data[index] = val
	this.length++
	return nil
}

//在数组尾部插入
func (this *Array) InsertTotail(val int) error {
	return this.Insert(this.Len(), val)
}

//删除数组指定索引
func (this *Array) Delete(index uint) (int, error) {
	if this.IsIndexOutOfRange(index) {
		return 0, errors.New("out of range")
	}

	val := this.data[index]
	//取出数组元素后交换下标元素值
	for i := index; i < this.Len()-1; i++ {
		this.data[i] = this.data[i+1]
	}
	this.length--
	return val, nil
}

func (this *Array) Print() {
	var format string
	for i := uint(0); i < this.Len(); i++ {
		format += fmt.Sprintf("|%+v", this.data[i])
	}
	fmt.Println(format)
}

File: php/05_array/array.php
<?php
/**
 * Created by PhpStorm.
 * User: duanfuhao
 * Date: 2019/8/21
 * Time: 11:09
 */

class MyArray {
    public $data;
    public $length;
    public $capacity;

    public function __construct($capacity)
    {
        $this->data = array();
        $this->capacity = $capacity; //数组最大容量
        $this->length = 0;  //数组现有数据长度
    }

    //判断数组是否已满
    public function isFull() {
        if ($this->length >= $this->capacity) {
            return true;
        }
        return false;
    }

    //判断下标是否越界
    public function isIndexOutOfRange($index) {
        if ($index < 0) {
            return true;
        }
        if ($index >= $this->capacity) { //下标是否越界
            return true;
        }
        return false;
    }

    //将数据插入索引指定位置
    public function insert($index,$val) {
        if ($this->isFull()) {
            throw new Exception("数组已满");
        }

        if ($this->isIndexOutOfRange($index)) {
            throw new Exception("下标越界 index:".$index);
        }
        //插入指定位置索引数据后，在此之后得索引如果有数据，需要交换
        for ($i = $this->length-1; $index <= $i; $i--) {
            $this->data[$i+1] = $this->data[$i];
        }
        $this->data[$index] = $val;
        $this->length++;
        return true;
    }

    //更加下标查找值
    public function find($index) {
        if ($this->isIndexOutOfRange($index)) {
            throw new Exception("下标越界 index:".$index);
        }
        return $this->data[$index];
    }

    //删除指定索引
    public function delete($index) {
        if ($this->isIndexOutOfRange($index)) {
            throw new Exception("下标越界 index:".$index);
        }
        //删除指定索引后，后面元素需要往前移动，保持数组连续性
        $val = $this->data[$index];
        for ($i = $index; $i < $this->length-1; $i++) {
            $this->data[$i] = $this->data[$i+1];
        }
        $this->length--;
        return $val;
    }

    //打印数组
    public function printData() {
        $format = "";
        for ($i = 0; $i < $this->length; $i++) {
            $format .= "|" . $this->data[$i];
        }
        print($format . "\n");
    }
}